### PR TITLE
Improve code actions performance by loading the document on demand

### PR DESCRIPTION
**Pull request by rubenporras**
only if there are fix methods for the given diagnostic. This might saves
the resource load operation, which can be expensive.

### CODE REVIEW

File: org.eclipse.xtext.ide.tests/testlang-src/org/eclipse/xtext/ide/tests/testlanguage/ide/server/CodeActionService.java [hunk 1]
```diff
@@ -50,10 +53,22 @@ public List<Either<Command, CodeAction>> getCodeActions(ICodeActionService2.Opti
 		List<Either<Command, CodeAction>> actions = super.getCodeActions(options);
 		for (Diagnostic d : options.getCodeActionParams().getContext().getDiagnostics()) {
 			Object code = d.getCode().get();
-			if (TestLanguageValidator.INVALID_NAME.equals(code)) {
-				actions.add(Either.forLeft(fixInvalidName(d, options)));
-			} else if (TestLanguageValidator.UNSORTED_MEMBERS.equals(code)) {
-				actions.add(Either.forRight(fixUnsortedMembers(d, options)));
+			try {
```
Comment (szarnekow) [inline org.eclipse.xtext.ide.tests/testlang-src/org/eclipse/xtext/ide/tests/testlanguage/ide/server/CodeActionService.java:56]: Can we leave this test-lang untouched to see whether the former way to write fixes still works?

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/IQuickFixProvider.java [hunk 2]
```diff
@@ -33,10 +36,22 @@ public interface IQuickFixProvider {
...
+	 * @param fixMethods
+	 *            a list of methods that create the resolution
 	 * @return 0..n resolutions for the given issue
 	 */
 	List<DiagnosticResolution> getResolutions(Options options, Diagnostic diagnostic);
 
+	
+	/**
+	 *
+	 * Collects all of methods that create the resolution for the given diagnostic
```
Comment (szarnekow) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/IQuickFixProvider.java:48]: The JavaDoc doesn't match the implementation.

Is it possible to keep the original interface or make this a default method. Should be fine to always return true by default, shouldn't it?

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/IQuickFixProvider.java [hunk 3]
```diff
@@ -33,10 +36,22 @@ public interface IQuickFixProvider {
 	 *
 	 * @param options
 	 *            Contextual action options
-	 * @param diagnostic
-	 *            the diagnostic
+	 * @param fixMethods
```
Comment (szarnekow) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/IQuickFixProvider.java:39]: JavaDoc mismatch.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/ILanguageServerAccess.java [hunk 4]
```diff
@@ -33,15 +33,15 @@
  */
 public interface ILanguageServerAccess {
 	class Context {
-		private final Resource resource;
+		private final XtextResource resource;
```
Comment (szarnekow) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/ILanguageServerAccess.java:36]: Why did you change this?

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/codeActions/QuickFixCodeActionService.java [hunk 5]
```diff
@@ -56,11 +58,23 @@ public List<Either<Command, CodeAction>> getCodeActions(Options options) {
...
+								options.setResource(context.getResource());
+
+								List<Either<Command, CodeAction>> codeActions = new ArrayList<>();
+								quickfixes.getResolutions(options, diagnostic).stream()
+										.sorted(Comparator
+												.nullsLast(Comparator.comparing(DiagnosticResolution::getLabel)))
+										.forEach(r -> codeActions.add(Either.forRight(createFix(r, diagnostic))));
+								return codeActions;
+							}).get());
+				} catch (InterruptedException | ExecutionException e) {
```
Comment (szarnekow) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/codeActions/QuickFixCodeActionService.java:75]: I understand why it was done like this, but having the blocking call here looks wrong to me.
The approach might need some more thought.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/IQuickFixProvider.java (hunk 3 above)
Comment (rubenporras) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/IQuickFixProvider.java]: done.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/ILanguageServerAccess.java (hunk 4 above)
Comment (rubenporras) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/ILanguageServerAccess.java]: because org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/DiagnosticResolution.java had a variable declared with XtextResource, I have downgraded it to a Resource in the last version so that this is not needed.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/IQuickFixProvider.java (hunk 2 above)
Comment (rubenporras) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/editor/quickfix/IQuickFixProvider.java]: yes, it should be possible. done.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/codeActions/QuickFixCodeActionService.java (hunk 5 above)
Comment (rubenporras) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/codeActions/QuickFixCodeActionService.java:69]: we could avoid the blocking call if we can expose the workspace manage in the ILanguageServerAccess. What about that?

**Review by rubenporras — COMMENTED** [review]
Thanks for the review.

File: org.eclipse.xtext.ide.tests/testlang-src/org/eclipse/xtext/ide/tests/testlanguage/ide/server/CodeActionService.java (hunk 1 above)
Comment (rubenporras) [reply org.eclipse.xtext.ide.tests/testlang-src/org/eclipse/xtext/ide/tests/testlanguage/ide/server/CodeActionService.java]: not quite because someone needs to create the Resource and the Document. I have submited a newer version where this is abstracted, but I still needed to change the signature of the method a little bit.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/codeActions/QuickFixCodeActionService.java (hunk 5 above)
Comment (rubenporras) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/codeActions/QuickFixCodeActionService.java:69]: I have added a new method doSyncRead which fixes the problem, I hope you like the solution.

**Comment by rubenporras** [comment]
@szarnekow , what do you think of the change now? there is no longer a blocking call and the test looks almost as before (one just need to override one method to return true indicating that any QF is handled)

**Review by szarnekow — APPROVED** [review]

**Comment by rubenporras** [comment]
thanks, I will rebase and squash.
